network/manager.go

```diff
@@ -0,0 +1,302 @@
+		manager.allowedNetworks[name] = true
+	}
+
+	if manager.isMultiNetwork() {
+		// Get list of existing networks
+		result, err := manager.sm.WatchNetworks(ctx, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, n := range result.Snapshot {
+			if manager.isNetAllowed(n) {
+				manager.networks[n] = NewNetwork(sm, n, manager.ipMasq)
+			}
+		}
+	} else {
+		manager.networks[""] = NewNetwork(sm, "", manager.ipMasq)
+	}
+
+	return manager, nil
+}
+
+func lookupExtIface(ifname string) (*net.Interface, net.IP, error) {
+	var iface *net.Interface
+	var iaddr net.IP
+	var err error
+
+	if len(ifname) > 0 {
+		if iaddr = net.ParseIP(ifname); iaddr != nil {
+			iface, err = ip.GetInterfaceByIP(iaddr)
+			if err != nil {
+				return nil, nil, fmt.Errorf("Error looking up interface %s: %s", ifname, err)
+			}
+		} else {
+			iface, err = net.InterfaceByName(ifname)
+			if err != nil {
+				return nil, nil, fmt.Errorf("Error looking up interface %s: %s", ifname, err)
+			}
+		}
+	} else {
+		log.Info("Determining IP address of default interface")
+		if iface, err = ip.GetDefaultGatewayIface(); err != nil {
+			return nil, nil, fmt.Errorf("Failed to get default interface: %s", err)
+		}
+	}
+
+	if iaddr == nil {
+		iaddr, err = ip.GetIfaceIP4Addr(iface)
+		if err != nil {
+			return nil, nil, fmt.Errorf("Failed to find IPv4 address for interface %s", iface.Name)
+		}
+	}
+
+	return iface, iaddr, nil
+}
+
+func writeSubnetFile(path string, nw ip.IP4Net, ipMasq bool, sd *backend.SubnetDef) error {
+	dir, name := filepath.Split(path)
+	os.MkdirAll(dir, 0755)
+
+	tempFile := filepath.Join(dir, "."+name)
+	f, err := os.Create(tempFile)
+	if err != nil {
+		return err
+	}
+
+	// Write out the first usable IP by incrementing
+	// sn.IP by one
+	sn := sd.Lease.Subnet
+	sn.IP += 1
+
+	fmt.Fprintf(f, "FLANNEL_NETWORK=%s\n", nw)
+	fmt.Fprintf(f, "FLANNEL_SUBNET=%s\n", sn)
+	fmt.Fprintf(f, "FLANNEL_MTU=%d\n", sd.MTU)
+	_, err = fmt.Fprintf(f, "FLANNEL_IPMASQ=%v\n", ipMasq)
+	f.Close()
+	if err != nil {
+		return err
+	}
+
+	// rename(2) the temporary file to the desired location so that it becomes
+	// atomically visible with the contents
+	return os.Rename(tempFile, path)
+}
+
+func (m *Manager) RunNetwork(net *Network) {
+	sn := net.Init(m.extIface, m.iaddr, m.eaddr)
+	if sn != nil {
+		if m.isMultiNetwork() {
+			path := filepath.Join(opts.subnetDir, net.Name) + ".env"
+			if err := writeSubnetFile(path, net.Config.Network, m.ipMasq, sn); err != nil {
+				log.Warningf("%v failed to write subnet file: %s", net.Name, err)
+				return
+			}
+		} else {
+			if err := writeSubnetFile(opts.subnetFile, net.Config.Network, m.ipMasq, sn); err != nil {
+				log.Warningf("%v failed to write subnet file: %s", net.Name, err)
+				return
+			}
+			daemon.SdNotify("READY=1")
+		}
+
+				log.Warningf("Running network %v", net.Name)
+		net.Run()
+		log.Infof("%v exited", net.Name)
+	}
+}
+
+func (m *Manager) watchNetworks(ctx context.Context, wg sync.WaitGroup) {
```

Comment:
passing `WaitGroup` by value actually copies it and I think will cause trouble. Just create a new `WaitGroup` here and wait for it before returning from this function.